ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceCallMany.java

```diff
@@ -0,0 +1,180 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TraceCallMany extends TraceCall implements JsonRpcMethod {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TraceCallMany.class);
+  private static final ObjectMapper MAPPER_IGNORE_REVERT_REASON = new ObjectMapper();
+
+  public TraceCallMany(
+      final BlockchainQueries blockchainQueries,
+      final ProtocolSchedule protocolSchedule,
+      final TransactionSimulator transactionSimulator) {
+    super(blockchainQueries, protocolSchedule, transactionSimulator);
+
+    // The trace_call specification does not output the revert reason, so we have to remove it
+    MAPPER_IGNORE_REVERT_REASON.addMixIn(FlatTrace.class, MixInIgnoreRevertReason.class);
+  }
+
+  @Override
+  public String getName() {
+    return transactionSimulator != null ? RpcMethod.TRACE_CALL_MANY.getMethodName() : null;
+  }
+
+  @Override
+  protected BlockParameter blockParameter(final JsonRpcRequestContext request) {
+    final Optional<BlockParameter> maybeBlockParameter =
+        request.getOptionalParameter(2, BlockParameter.class);
+
+    if (maybeBlockParameter.isPresent()) {
+      return maybeBlockParameter.get();
+    }
+
+    return BlockParameter.LATEST;
+  }
+
+  @Override
+  protected Object resultByBlockNumber(
+      final JsonRpcRequestContext requestContext, final long blockNumber) {
+
+    if (requestContext.getRequest().getParamLength() != 2) {
+      return new JsonRpcErrorResponse(
+          requestContext.getRequest().getId(), JsonRpcError.INVALID_PARAMS);
+    }
+
+    final TraceCallManyParameter[] transactionsAndTraceTypeParameters;
+    try {
+      transactionsAndTraceTypeParameters =
+          requestContext.getRequiredParameter(0, TraceCallManyParameter[].class);
+    } catch (final Exception e) {
+      LOG.error("Error parsing trace call many parameter: {}", e.getLocalizedMessage());
+      return new JsonRpcErrorResponse(
+          requestContext.getRequest().getId(), JsonRpcError.INVALID_PARAMS);
+    }
+
+    final Optional<BlockHeader> maybeBlockHeader =
+        blockchainQueries.get().getBlockHeaderByNumber(blockNumber);
+
+    if (maybeBlockHeader.isEmpty()) {
+      return new JsonRpcErrorResponse(requestContext.getRequest().getId(), BLOCK_NOT_FOUND);
+    }
+
+    final List<JsonNode> traceCallResults = new ArrayList<>();
+
+    final WorldUpdater updater = transactionSimulator.getWorldUpdater(maybeBlockHeader.get());
+    try {
+      Arrays.stream(transactionsAndTraceTypeParameters)
+          .forEachOrdered(
+              param -> {
+                final WorldUpdater finalUpdater = updater.updater();
+                try {
+                  executeSingleCall(
+                      param.getTuple().getJsonCallParameter(),
+                      param.getTuple().getTraceTypeParameter(),
+                      maybeBlockHeader.get(),
+                      finalUpdater,
+                      traceCallResults);
```

Comment:
I think it would be cleaner to have this method return a single result and add it to the traceCallResults collection here rather than passing it into the executeSingleCall method

Reply:
[c4f0086](https://github.com/hyperledger/besu/pull/3468/commits/c4f008654f1582a71db15a3d2d18de30fed3949b)